### ISSUE TITLE
Can turn immutable into mut with `ref mut`

### ISSUE THREAD

**Issue by Esption** [rust-lang/rust#52240]
This lets you get a mutable reference to an immutable value under a specific case.

Code:
```Rust
enum Foo {
	Bar(i32),
}

fn main() {
	let arr = vec!(Foo::Bar(0));
	if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
		*val = 9001;
	}
	match arr[0] {
		Foo::Bar(ref s) => println!("{}", s)
	}
}
```

From what I can tell, the tuple and at least two depths of Enums are necessary, along with an `&` on the right side.  There's probably other variations, but that seems to work well enough.

Expected output would be a compile error, instead output is `9001`

This works on both stable `rustc 1.27.1 (5f2b325f6 2018-07-07)` and nightly `rustc 1.29.0-nightly (e5f6498d3 2018-07-10)`

**Comment by sfackler** [comment]
@arielb1 would #52232 fix this as well?

**Comment by hellow554** [comment]
Fixed by `#![feature(nll)]`. 

Off: I wonder why we can't ship nll yet. It would solve so many issues right now, because of the match ergonomics *sigh*
Match ergonomics produces roughly 10% of all errors/unsoundness/ICEs atm (at least it feels like this)

**Comment by kennytm** [comment]
Yes #52232 correctly fixed this issue.

```console
$ rustup-toolchain-install-master 4700e1188f66fdb4086b7593416b678b8fe935f4
downloading <https://s3-us-west-1.amazonaws.com/rust-lang-ci2/rustc-builds/4700e1188f66fdb4086b7593416b678b8fe935f4/rustc-nightly-x86_64-apple-darwin.tar.xz>...
completed                                                                                                                                                                                                                                  
downloading <https://s3-us-west-1.amazonaws.com/rust-lang-ci2/rustc-builds/4700e1188f66fdb4086b7593416b678b8fe935f4/rust-std-nightly-x86_64-apple-darwin.tar.xz>...
completed                                                                                                                                                                                                                                  
toolchain `4700e1188f66fdb4086b7593416b678b8fe935f4` is successfully installed!

$ rustc +4700e1188f66fdb4086b7593416b678b8fe935f4 1.rs
error[E0596]: cannot borrow field of immutable binding as mutable
 --> 1.rs:7:24
  |
7 |     if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
  |                           ^^^^^^^^^^^ cannot mutably borrow field of immutable binding

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
```
